File: src/Map/lc0001_TwoSum.go
package Map

import "github.com/k0kubun/pp/v3"

// https://leetcode.cn/problems/two-sum
func twoSum(nums []int, target int) []int {
	numToIndexMap := make(map[int]int, len(nums))
	for i, num := range nums {
		if index, ok := numToIndexMap[target-num]; ok {
			return []int{index, i}
		}
		numToIndexMap[num] = i
	}
	return []int{}
}

func TestTwoSum() {
	pp.Print(twoSum([]int{2, 7, 11, 15}, 9))
}

File: src/Map/lc0001_TwoSum.py
# https://leetcode.cn/problems/two-sum
def two_sum(nums: list[int], target: int) -> tuple[int, int]:
    num_to_index_dict = {}
    for i, num in enumerate(nums):
        # dict中所有的value都是int，所以不存在value为None的key-value pair
        index = num_to_index_dict.get(target - num)
        if index is not None:
            return index, i
        num_to_index_dict[num] = i
    raise "unexpected input"
